src/braket/emulation/device_emulator_properties.py

```diff
@@ -0,0 +1,232 @@
+        twoQubitProperties = values["twoQubitProperties"]
+        oneQubitProperties = values.get("oneQubitProperties")
+        indices = list(oneQubitProperties.keys())
+        qubit_labels = sorted(int(x) for x in indices)
+
+        for edge in twoQubitProperties:
+            node_1, node_2 = edge.split("-")
+            cls.node_validator(node_1, qubit_labels, "twoQubitProperties")
+            cls.node_validator(node_2, qubit_labels, "twoQubitProperties")
+
+        return values
+
+    @validator("supportedResultTypes", pre=False)
+    @classmethod
+    def validate_supportedResultTypes(cls, supportedResultTypes: list) -> list:
+        valid_result_types = [rt.name for rt in DEFAULT_SUPPORTED_RESULT_TYPES]
+
+        for result_type in supportedResultTypes:
+            # Check if result type is one of the valid types
+            if result_type.name not in valid_result_types:
+                raise ValueError(
+                    f"Invalid result type. Must be one of: {', '.join(valid_result_types)}"
+                )
+        return supportedResultTypes
+
+    @property
+    def qubit_labels(self) -> list:
+        indices = list(self.oneQubitProperties.keys())
+        return sorted(int(x) for x in indices)
+
+    @property
+    def fully_connected(self) -> bool:
+        """Determine if the connectivity graph is fully connected.
+
+        Note: We treat the graph as undirected, and determine if it is
+            a complete graph by counting the number of distinct edges
+        """
+        if not self.connectivityGraph:
+            return True
+
+        edges = set()
+        for node, neighbors in self.connectivityGraph.items():
+            edges_node = [(int(node), int(neighbor)) for neighbor in neighbors]
+            edges_node = [(min(edge), max(edge)) for edge in edges_node]
+            edges.update(edges_node)
+
+        return len(edges) == self.qubitCount * (self.qubitCount - 1) / 2
+
+    @property
+    def directed(self) -> bool:
+        """Determine if the connectivity graph is a directed graph."""
+        for node, neighbors in self.connectivityGraph.items():
+            for neighbor in neighbors:
+                # If neighbor doesn't link back to node, it's directed
+                if node not in self.connectivityGraph.get(neighbor, []):
+                    return True
+        return False
+
+    @classmethod
+    def from_device_properties(
+        cls, device_properties: DeviceCapabilities
+    ) -> "DeviceEmulatorProperties":
+        if isinstance(device_properties, IonqDeviceCapabilities):
+            device_properties = standardize_ionq_device_properties(device_properties)
+        if isinstance(device_properties, DeviceCapabilities):
+            return cls._from_json(device_properties.json())
+        raise ValueError("device_properties has to be an instance of DeviceCapabilities.")
+
+    @classmethod
+    def _from_json(cls, device_properties_json: str) -> "DeviceEmulatorProperties":
```

Comment:
Why use this instead of constructing directly from the object? Relying on JSON kvps isn't as safe.